pkg/modulewriter/tfwriter.go

```diff
@@ -166,27 +166,49 @@ func writeMain(
 
 	for _, mod := range modules {
 		hclBody.AppendNewline()
-		// Add block
-		moduleBlock := hclBody.AppendNewBlock("module", []string{string(mod.ID)})
-		moduleBody := moduleBlock.Body()
-
-		// Add source attribute
-		ds, err := DeploymentSource(mod)
-		if err != nil {
+		if err := writeModule(mod, hclBody); err != nil {
 			return err
 		}
-		moduleBody.SetAttributeValue("source", cty.StringVal(ds))
-
-		// For each Setting
-		for _, setting := range orderKeys(mod.Settings.Items()) {
-			value := mod.Settings.Get(setting)
-			moduleBody.SetAttributeRaw(setting, config.TokensForValue(value))
-		}
 	}
 
 	return writeHclFile(filepath.Join(dst, "main.tf"), hclFile)
 }
 
+func writeModule(mod config.Module, b *hclwrite.Body) error {
+	mb := b.AppendNewBlock("module", []string{string(mod.ID)}).Body()
+
+	ds, err := DeploymentSource(mod)
+	if err != nil {
+		return err
+	}
+	mb.SetAttributeValue("source", cty.StringVal(ds))
+
+	for _, setting := range orderKeys(mod.Settings.Items()) {
+		value := mod.Settings.Get(setting)
+		mb.SetAttributeRaw(setting, config.TokensForValue(value))
+	}
+
+	if mod.Protected {
+		rn := fmt.Sprintf("protection_%s", mod.ID)
+		rb := b.AppendNewBlock("resource", []string{"null_resource", rn}).Body()
+
+		mr := cty.StringVal(fmt.Sprintf("module.%s", mod.ID))
+		// TODO: Warning: Quoted references are deprecated
+		rb.SetAttributeValue("depends_on", cty.TupleVal([]cty.Value{mr}))
+
+		// tb := rb.AppendNewBlock("triggers", []string{}).Body()
+		// for _, setting := range orderKeys(mod.Settings.Items()) {
+		// 	value := mod.Settings.Get(setting)
+		// 	tb.SetAttributeRaw(setting, config.TokensForValue(value))
+		// }
+		rb.SetAttributeRaw("triggers", config.TokensForValue(mod.Settings.AsObject()))
```

Comment:
In your example, this renders as "deployment_name", "project_id" and "region". So if any of these change, it would trigger a destruction, is that correct?

Reply:
Yes, we lock all input values on protected module